runtime/compiler/runtime/SymbolValidationManager.cpp

```diff
@@ -232,388 +211,300 @@ TR::SymbolValidationManager::storeRecord(void *symbol, TR::SymbolValidationRecor
-      return false;
-   if (inHeuristicRegion())
-      return true;
-
-   int32_t arrayDims = 0;
-   clazz = getBaseComponentClass(clazz, arrayDims);
-
    TR_OpaqueClassBlock *beholder = reinterpret_cast<TR_OpaqueClassBlock *>(J9_CLASS_FROM_CP(constantPoolOfBeholder));
    SVM_ASSERT_ALREADY_VALIDATED(this, beholder);
-
-   SymbolValidationRecord *record = new (_region) StaticClassFromCPRecord(clazz, beholder, cpIndex);
-   return storeValidationRecordIfNecessary(static_cast<void *>(clazz), record, arrayDims);
+   return addClassRecord(clazz, new (_region) StaticClassFromCPRecord(clazz, beholder, cpIndex));
    }
 
 bool
 TR::SymbolValidationManager::addClassFromMethodRecord(TR_OpaqueClassBlock *clazz, TR_OpaqueMethodBlock *method)
    {
-   if (!clazz)
-      return false;
-   if (inHeuristicRegion())
-      return true;
-
-   int32_t arrayDims = 0;
-   clazz = getBaseComponentClass(clazz, arrayDims);
-
    SVM_ASSERT_ALREADY_VALIDATED(this, method);
-
-   SymbolValidationRecord *record = new (_region) ClassFromMethodRecord(clazz, method);
-   return storeValidationRecordIfNecessary(static_cast<void *>(clazz), record);
+   return addClassRecord(clazz, new (_region) ClassFromMethodRecord(clazz, method));
    }
 
 bool
 TR::SymbolValidationManager::addComponentClassFromArrayClassRecord(TR_OpaqueClassBlock *componentClass, TR_OpaqueClassBlock *arrayClass)
    {
-   if (!componentClass)
-      return false;
-   if (inHeuristicRegion())
-      return true;
-
+   // Class chain validation for the base component type is already taken care of
    SVM_ASSERT_ALREADY_VALIDATED(this, arrayClass);
-
-   SymbolValidationRecord *record = new (_region) ComponentClassFromArrayClassRecord(componentClass, arrayClass);
-   return storeValidationRecordIfNecessary(static_cast<void *>(componentClass), record);
+   return addVanillaRecord(componentClass, new (_region) ComponentClassFromArrayClassRecord(componentClass, arrayClass));
    }
 
 bool
 TR::SymbolValidationManager::addArrayClassFromComponentClassRecord(TR_OpaqueClassBlock *arrayClass, TR_OpaqueClassBlock *componentClass)
    {
-   if (!arrayClass)
-      return false;
-   if (inHeuristicRegion())
-      return true;
-
+   // Class chain validation for the base component type is already taken care of
    SVM_ASSERT_ALREADY_VALIDATED(this, componentClass);
-
-   SymbolValidationRecord *record = new (_region) ArrayClassFromComponentClassRecord(arrayClass, componentClass);
-   return storeValidationRecordIfNecessary(static_cast<void *>(arrayClass), record);
+   return addVanillaRecord(arrayClass, new (_region) ArrayClassFromComponentClassRecord(arrayClass, componentClass));
    }
 
 bool
 TR::SymbolValidationManager::addSuperClassFromClassRecord(TR_OpaqueClassBlock *superClass, TR_OpaqueClassBlock *childClass)
    {
-   if (!superClass)
-      return false;
-   if (inHeuristicRegion())
-      return true;
-
-   int32_t arrayDims = 0;
-   superClass = getBaseComponentClass(superClass, arrayDims);
```

Comment:
For my own clarification, it is impossible for a superclass of some class to be an array class?